### PR TITLE
Use the custom memcpy for realloc.

### LINKED ISSUE #154: Improve SIMD memcpy on windows?
I saw an interesting issue at mimalloc: https://github.com/microsoft/mimalloc/issues/201.

It seems that on Windows memcpy does not invoke SIMD instructions by default. I remember that musl [memcpy](https://github.com/esmil/musl/blob/master/src/string/memcpy.c) does not use SIMD directly either, though it is written in a way that makes the compiler feel happy to optimize.

There are three cases of memcpy in snmalloc. Therefore, perhaps writing memcpy using SIMD instructions directly on some platforms may result in a better performance.

### PR DESCRIPTION
**Pull request by davidchisnall**
This wraps our memcpy with some assumptions that let the optimiser know
that we're copying chunks that are strongly aligned.  With clang 13 on
x86, this generates three variants:

 - A special case for 16 bytes that's a single vector load + store.
 - A vector-copy loop for sizes <512 bytes.
 - rep movsb for larger sizes.

This is almost certainly faster than the platform memcpy (if for no
other reason than that it doesn't have to care about handling unaligned
copies).

I don't know if it will show up in benchmarks, but if it does then it
Fixes #154

### CODE REVIEW

**Comment by davidchisnall** [comment]
@nwf, the memcpy is currently incorrect for CHERI.  It's probably worth tweaking the default one for any platform where the AAL says that it's CHERI and a specialising it for Morello...

File: src/override/malloc.cc [hunk 1]
```diff
@@ -9,6 +11,31 @@ using namespace snmalloc;
...
+   * object size and that the start and end are strongly aligned.
+   */
+  SNMALLOC_FAST_PATH
+  void memcpy_for_realloc(void* dst, const void* src, size_t size)
+  {
+    // Tell the compiler some things that it can use to optimise the memcpy:
+    // The size is at least the minimum alloc size (skip all of the small copy
+    // things) and the size is a multiple of the smallest size.
+    SNMALLOC_ASSUME(size >= MIN_ALLOC_SIZE);
+    SNMALLOC_ASSUME((size % MIN_ALLOC_SIZE) == 0);
```
[the remark this answers is not in the document]
Comment (davidchisnall) [reply src/override/malloc.cc:27]: I think that was fixed prior to the comment?

File: src/override/malloc.cc [hunk 2]
```diff
@@ -91,7 +118,7 @@ extern "C"
       // Guard memcpy as GCC is assuming not nullptr for ptr after the memcpy
       // otherwise.
       if (sz != 0)
-        memcpy(p, ptr, sz);
+        memcpy_for_realloc(p, ptr, bits::max(MIN_ALLOC_SIZE, sz));
```
Comment (mjp41) [inline src/override/malloc.cc:121]: Will `max` compile to a branch.  Perhaps, do a round up, which is branch free.

File: src/override/malloc.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjp41) [reply src/override/malloc.cc:27]: Yeah, stale page.  But I don't think it is quite fixed. 
```
auto a = malloc(32);
auto b = realloc(a, 17);
```

[the remark this answers is not in the document]
Comment (davidchisnall) [reply src/override/malloc.cc:27]: Fixed!

**Comment by mjp41** [comment]
So running some of my usual benchmarks does not show any statistically significant difference.   I am still happy to take this. I wonder if it makes sense to have a micro-benchmark to test a collection of reallocs to see if this is a win in an artificial scenario.

**Comment by davidchisnall** [comment]
If it isn't a bottleneck then it's probably not worth the code size increase.  It's small but a lot of small things add up.

**Comment by mjp41** [comment]
> If it isn't a bottleneck then it's probably not worth the code size increase. It's small but a lot of small things add up.

I'll close, we can always revisit.

**Pull request closed without merging**